File: typescript/src/resources/captcha.ts
/** Captcha resources (`/captcha/*`). */

import { Transport } from '../runtime/transport';

class Recognition {
  constructor(private transport: Transport) {}

  hcaptcha(opts: {
    queries?: string[];
    question?: string;
    async?: boolean;
    [key: string]: unknown;
  } = {}): Promise<Record<string, unknown>> {
    const { queries, question, ...rest } = opts;
    const body: Record<string, unknown> = { ...rest };
    if (queries !== undefined) body.queries = queries;
    if (question !== undefined) body.question = question;
    return this.transport.request('POST', '/captcha/recognition/hcaptcha', { json: body });
  }

  image2text(opts: {
    image: string;
    async?: boolean;
    [key: string]: unknown;
  }): Promise<Record<string, unknown>> {
    const { image, ...rest } = opts;
    const body: Record<string, unknown> = { image, ...rest };
    return this.transport.request('POST', '/captcha/recognition/image2text', { json: body });
  }

  recaptcha2(opts: {
    image: string;
    question: string;
    async?: boolean;
    [key: string]: unknown;
  }): Promise<Record<string, unknown>> {
    const { image, question, ...rest } = opts;
    const body: Record<string, unknown> = { image, question, ...rest };
    return this.transport.request('POST', '/captcha/recognition/recaptcha2', { json: body });
  }
}

class Token {
  constructor(private transport: Transport) {}

  hcaptcha(opts: {
    websiteKey: string;
    websiteUrl: string;
    rqdata?: string;
    proxy?: string;
    async?: boolean;
    [key: string]: unknown;
  }): Promise<Record<string, unknown>> {
    const { websiteKey, websiteUrl, rqdata, proxy, ...rest } = opts;
    const body: Record<string, unknown> = {
      website_key: websiteKey,
      website_url: websiteUrl,
      ...rest,
    };
    if (rqdata !== undefined) body.rqdata = rqdata;
    if (proxy !== undefined) body.proxy = proxy;
    return this.transport.request('POST', '/captcha/token/hcaptcha', { json: body });
  }

  recaptcha2(opts: {
    websiteKey: string;
    websiteUrl: string;
    proxy?: string;
    async?: boolean;
    [key: string]: unknown;
  }): Promise<Record<string, unknown>> {
    const { websiteKey, websiteUrl, proxy, ...rest } = opts;
    const body: Record<string, unknown> = {
      website_key: websiteKey,
      website_url: websiteUrl,
      ...rest,
    };
    if (proxy !== undefined) body.proxy = proxy;
    return this.transport.request('POST', '/captcha/token/recaptcha2', { json: body });
  }

  recaptcha3(opts: {
    pageAction: string;
    websiteKey: string;
    websiteUrl: string;
    async?: boolean;
    [key: string]: unknown;
  }): Promise<Record<string, unknown>> {
    const { pageAction, websiteKey, websiteUrl, ...rest } = opts;
    const body: Record<string, unknown> = {
      page_action: pageAction,
      website_key: websiteKey,
      website_url: websiteUrl,
      ...rest,
    };
    return this.transport.request('POST', '/captcha/token/recaptcha3', { json: body });
  }

  turnstile(opts: {
    websiteKey: string;
    websiteUrl: string;
    action?: string;
    cdata?: string;
    async?: boolean;
    [key: string]: unknown;
  }): Promise<Record<string, unknown>> {
    const { websiteKey, websiteUrl, action, cdata, ...rest } = opts;
    const body: Record<string, unknown> = {
      website_key: websiteKey,
      website_url: websiteUrl,
      ...rest,
    };
    if (action !== undefined) body.action = action;
    if (cdata !== undefined) body.cdata = cdata;
    return this.transport.request('POST', '/captcha/token/turnstile', { json: body });
  }
}

class Tasks {
  constructor(private transport: Transport) {}

  retrieve(opts: { taskId: string; [key: string]: unknown }): Promise<Record<string, unknown>> {
    const { taskId, ...rest } = opts;
    return this.transport.request('POST', '/captcha/tasks', {
      json: { task_id: taskId, ...rest },
    });
  }
}

export class Captcha {
  readonly recognition: Recognition;
  readonly token: Token;
  readonly tasks: Tasks;

  constructor(transport: Transport) {
    this.recognition = new Recognition(transport);
    this.token = new Token(transport);
    this.tasks = new Tasks(transport);
  }
}

File: python/src/acedatacloud/resources/captcha.py
"""Captcha resources (``/captcha/*``)."""

from __future__ import annotations

from typing import Any


class _Recognition:
    def __init__(self, transport: Any) -> None:
        self._transport = transport

    def hcaptcha(
        self,
        *,
        queries: list[str] | None = None,
        question: str | None = None,
        async_: bool | None = None,
        **kwargs: Any,
    ) -> dict[str, Any]:
        body: dict[str, Any] = {**kwargs}
        if queries is not None:
            body["queries"] = queries
        if question is not None:
            body["question"] = question
        if async_ is not None:
            body["async"] = async_
        return self._transport.request("POST", "/captcha/recognition/hcaptcha", json=body)

    def recaptcha2(
        self,
        *,
        image: str,
        question: str,
        async_: bool | None = None,
        **kwargs: Any,
    ) -> dict[str, Any]:
        body: dict[str, Any] = {"image": image, "question": question, **kwargs}
        if async_ is not None:
            body["async"] = async_
        return self._transport.request("POST", "/captcha/recognition/recaptcha2", json=body)


class _Token:
    def __init__(self, transport: Any) -> None:
        self._transport = transport

    def hcaptcha(
        self,
        *,
        website_key: str,
        website_url: str,
        rqdata: str | None = None,
        proxy: str | None = None,
        async_: bool | None = None,
        **kwargs: Any,
    ) -> dict[str, Any]:
        body: dict[str, Any] = {
            "website_key": website_key,
            "website_url": website_url,
            **kwargs,
        }
        if rqdata is not None:
            body["rqdata"] = rqdata
        if proxy is not None:
            body["proxy"] = proxy
        if async_ is not None:
            body["async"] = async_
        return self._transport.request("POST", "/captcha/token/hcaptcha", json=body)

    def recaptcha2(
        self,
        *,
        website_key: str,
        website_url: str,
        proxy: str | None = None,
        async_: bool | None = None,
        **kwargs: Any,
    ) -> dict[str, Any]:
        body: dict[str, Any] = {"website_key": website_key, "website_url": website_url, **kwargs}
        if proxy is not None:
            body["proxy"] = proxy
        if async_ is not None:
            body["async"] = async_
        return self._transport.request("POST", "/captcha/token/recaptcha2", json=body)

    def recaptcha3(
        self,
        *,
        website_key: str,
        website_url: str,
        page_action: str,
        async_: bool | None = None,
        **kwargs: Any,
    ) -> dict[str, Any]:
        body: dict[str, Any] = {
            "website_key": website_key,
            "website_url": website_url,
            "page_action": page_action,
            **kwargs,
        }
        if async_ is not None:
            body["async"] = async_
        return self._transport.request("POST", "/captcha/token/recaptcha3", json=body)


class _Tasks:
    def __init__(self, transport: Any) -> None:
        self._transport = transport

    def retrieve(self, *, task_id: str, **kwargs: Any) -> dict[str, Any]:
        return self._transport.request("POST", "/captcha/tasks", json={"task_id": task_id, **kwargs})


class Captcha:
    """Synchronous captcha client."""

    def __init__(self, transport: Any) -> None:
        self.recognition = _Recognition(transport)
        self.token = _Token(transport)
        self.tasks = _Tasks(transport)


class _AsyncRecognition:
    def __init__(self, transport: Any) -> None:
        self._transport = transport

    async def hcaptcha(
        self,
        *,
        queries: list[str] | None = None,
        question: str | None = None,
        async_: bool | None = None,
        **kwargs: Any,
    ) -> dict[str, Any]:
        body: dict[str, Any] = {**kwargs}
        if queries is not None:
            body["queries"] = queries
        if question is not None:
            body["question"] = question
        if async_ is not None:
            body["async"] = async_
        return await self._transport.request("POST", "/captcha/recognition/hcaptcha", json=body)

    async def recaptcha2(
        self,
        *,
        image: str,
        question: str,
        async_: bool | None = None,
        **kwargs: Any,
    ) -> dict[str, Any]:
        body: dict[str, Any] = {"image": image, "question": question, **kwargs}
        if async_ is not None:
            body["async"] = async_
        return await self._transport.request("POST", "/captcha/recognition/recaptcha2", json=body)


class _AsyncToken:
    def __init__(self, transport: Any) -> None:
        self._transport = transport

    async def hcaptcha(
        self,
        *,
        website_key: str,
        website_url: str,
        rqdata: str | None = None,
        proxy: str | None = None,
        async_: bool | None = None,
        **kwargs: Any,
    ) -> dict[str, Any]:
        body: dict[str, Any] = {
            "website_key": website_key,
            "website_url": website_url,
            **kwargs,
        }
        if rqdata is not None:
            body["rqdata"] = rqdata
        if proxy is not None:
            body["proxy"] = proxy
        if async_ is not None:
            body["async"] = async_
        return await self._transport.request("POST", "/captcha/token/hcaptcha", json=body)

    async def recaptcha2(
        self,
        *,
        website_key: str,
        website_url: str,
        proxy: str | None = None,
        async_: bool | None = None,
        **kwargs: Any,
    ) -> dict[str, Any]:
        body: dict[str, Any] = {"website_key": website_key, "website_url": website_url, **kwargs}
        if proxy is not None:
            body["proxy"] = proxy
        if async_ is not None:
            body["async"] = async_
        return await self._transport.request("POST", "/captcha/token/recaptcha2", json=body)

    async def recaptcha3(
        self,
        *,
        website_key: str,
        website_url: str,
        page_action: str,
        async_: bool | None = None,
        **kwargs: Any,
    ) -> dict[str, Any]:
        body: dict[str, Any] = {
            "website_key": website_key,
            "website_url": website_url,
            "page_action": page_action,
            **kwargs,
        }
        if async_ is not None:
            body["async"] = async_
        return await self._transport.request("POST", "/captcha/token/recaptcha3", json=body)


class _AsyncTasks:
    def __init__(self, transport: Any) -> None:
        self._transport = transport

    async def retrieve(self, *, task_id: str, **kwargs: Any) -> dict[str, Any]:
        return await self._transport.request("POST", "/captcha/tasks", json={"task_id": task_id, **kwargs})


class AsyncCaptcha:
    """Async captcha client."""

    def __init__(self, transport: Any) -> None:
        self.recognition = _AsyncRecognition(transport)
        self.token = _AsyncToken(transport)
        self.tasks = _AsyncTasks(transport)
